src/supervisor/models.py

```diff
@@ -0,0 +1,126 @@
+        Path to the logfile
+
+    Methods
+    -------
+    add_message_handler(handler)
+        Register a new handler for sending messages
+    notify(message)
+        Send a notification to all registered handlers
+    _manage_exceptions
+        Closure around a replacement for sys.excepthook
+    """
+
+    def __init__(self, project_name, log_path=None):
+
+        #: Set up our list of MessageHandlers and add a default ConsoleHandler
+        self.project_name = project_name
+        self.message_handlers = []
+        self.message_handlers.append(ConsoleHandler())
+        self.log_path = log_path
+
+        #: Catch any uncaught exception with our custom exception handler
+        sys.excepthook = self._manage_exceptions()
+
+    def add_message_handler(self, handler: MessageHandler):
+        """Register a new handler for sending messages
+
+        Parameters:
+        -----------
+        handler : MessageHandler
+            A MessageHandler instance for the desired notification method.
+        """
+        self.message_handlers.append(handler)
+
+    def notify(self, message):
+        """Send a notification to all registered handlers by calling .send_message() """
+        for handler in self.message_handlers:
+            handler.send_message(message)
+
+    def _manage_exceptions(self):
+        """Closure around a replacement for sys.excepthook
+
+        A closure ensures global_exception_handler() has access to self.notify to send notifications while still
+        maintaining the signature required by sys.excepthook
+        """
+
+        log = logging.getLogger(self.project_name)
+
+        def global_exception_handler(exc_class, exc_object, tb):  # pylint: disable=invalid-name
+            """Log any uncaught exceptions and send a notification via self.notify
+
+            Parameters
+            ----------
+            exc_class : Type
+                The type of the exception
+            exc_object : Exception
+                The exception object
+            tb : Traceback
+                Original traceback
+
+            Used to handle any uncaught exceptions. Formats an error message, logs it, and sends an email.
+            """
+
+            last_traceback = (traceback.extract_tb(tb))[-1]
+            line_number = last_traceback[1]
+            file_name = last_traceback[0].split('.')[0]
+            error = os.linesep.join(traceback.format_exception(exc_class, exc_object, tb))
+
+            log.error(f'global error handler line: {line_number} ({file_name})')  # pylint: disable=logging-fstring-interpolation
+            log.error(error)
+
+            message_details = MessageDetails()
+            message_details.message = error
+            message_details.subject = f'{self.project_name}: ERROR'
+            message_details.log_file = self.log_path
+            self.notify(message_details)
+
+        return global_exception_handler
+
+
+class MessageDetails:  # pylint: disable=too-few-public-methods
```

Comment:
**question**: Would this be a good use case for a named tuple? Maybe those wouldn't provide the type-checking that you're doing in the add_message_handler method?

Reply:
Yeah, my goal is to require the end user to use readme.md or other external documentation as little as possible so I made it its own class for type hinting and code completion. I want to take a stab at the Null-Object pattern at some point too.

Another option is to use the `dataclass` decorator introduced in python 3.7 once we've had a chance to get Pro 2.7 installed across the board: https://docs.python.org/3/library/dataclasses.html.